bindings/rust/s2n-tls/src/testing.rs

```diff
@@ -325,3 +335,185 @@ pub fn poll_tls_pair_result(pair: &mut Pair<Harness, Harness>) -> Result<()> {
+/// ```
+//
+// The doctest is `ignore`d because testing utilities are not publicly exported
+// and therefore can't be referenced in a doc comment.
+//
+// We allow dead_code, because otherwise the compiler complains about the tx_streams
+// never being read. This is because it can't reason through the pointers that were
+// passed into the s2n-tls connection io contexts.
+#[allow(dead_code)]
+pub struct TestPair {
+    pub server: connection::Connection,
+    pub client: connection::Connection,
+
+    // Pin: since we are dereferencing this pointer (because it is passed as the send/recv ctx)
+    // we need to ensure that the pointer remains in the same place
+    // Box: A Vec (or VecDeque) may be moved or reallocated, so we need another layer of
+    // indirection to have a stable (pinned) reference
+    /// a data buffer that the server writes to and the client reads from
+    server_tx_stream: Pin<Box<LocalDataBuffer>>,
+    /// a data buffer that the client writes to and the server reads from
+    client_tx_stream: Pin<Box<LocalDataBuffer>>,
+}
+
+impl TestPair {
+    /// utility method to test simple handshakes
+    ///
+    /// Create a client and server from the associated `config`, and try to complete
+    /// a TLS handshake. The result of the handshake is returned.
+    pub fn handshake_with_config(config: &config::Config) -> Result<(), error::Error> {
+        Self::from_configs(config, config).handshake()
+    }
+
+    /// create a pair from a config
+    ///
+    /// A server and client connection will be created, and both will be associated
+    /// with `config`. The connections will be setup for IO over shared memory,
+    /// but no IO is performed. To handshake the connections, call `handshake()`.
+    pub fn from_config(config: &config::Config) -> Self {
+        Self::from_configs(config, config)
+    }
+
+    pub fn from_configs(client_config: &config::Config, server_config: &config::Config) -> Self {
+        let client_tx_stream = Box::pin(Default::default());
+        let server_tx_stream = Box::pin(Default::default());
+
+        let client = Self::register_connection(
+            enums::Mode::Client,
+            client_config,
+            &client_tx_stream,
+            &server_tx_stream,
+        )
+        .unwrap();
+
+        let server = Self::register_connection(
+            enums::Mode::Server,
+            server_config,
+            &server_tx_stream,
+            &client_tx_stream,
+        )
+        .unwrap();
+
+        Self {
+            server,
+            client,
+            server_tx_stream,
+            client_tx_stream,
+        }
+    }
+
+    /// create a connection ready for harness IO
+    ///
+    /// This mostly consists of setting the IO callbacks and the IO contexts, but
+    /// we also set blinding to "SelfService" to avoid really slow unit tests. Note
+    /// that SelfService blinding does not completely disable blinding, and will
+    /// still affect "shutdown" workflows.
```

Comment:
This comment is better, but I still don't think it really does the job of warning users what unexpected behavior might happen. How about:
```suggestion
    /// This mostly consists of setting the IO callbacks and the IO contexts.
    ///
    /// We also set blinding to "SelfService" to avoid long delays after failures in unit tests.
    /// However, this will cause calls to `poll_shutdown` to return Poll::Pending until the blinding delay elapses.
```